scripts/add_electricity.py

```diff
@@ -301,25 +302,46 @@ def attach_wind_and_solar(n, costs, input_profiles, technologies, line_length_fa
                    p_max_pu=ds['profile'].transpose('time', 'bus').to_pandas())
 
 
-def attach_conventional_generators(n, costs, ppl, carriers):
+def attach_conventional_generators(n, costs, ppl, conventional_carriers, extendable_carriers, **config):
 
+    carriers = set(conventional_carriers) | set(extendable_carriers['Generator'])
     _add_missing_carriers_from_costs(n, costs, carriers)
 
-    ppl = (ppl.query('carrier in @carriers').join(costs, on='carrier')
+    ppl = (ppl.query('carrier in @carriers').join(costs, on='carrier', rsuffix='_r')
            .rename(index=lambda s: 'C' + str(s)))
+    ppl["efficiency"] = ppl.efficiency.fillna(ppl.efficiency_r)
 
-    logger.info('Adding {} generators with capacities [MW] \n{}'
-                .format(len(ppl), ppl.groupby('carrier').p_nom.sum()))
+    logger.info('Adding {} generators with capacities [GW] \n{}'
+                .format(len(ppl), ppl.groupby('carrier').p_nom.sum().div(1e3).round(2)))
 
     n.madd("Generator", ppl.index,
            carrier=ppl.carrier,
            bus=ppl.bus,
-           p_nom=ppl.p_nom,
+           p_nom_min=ppl.p_nom.where(ppl.carrier.isin(conventional_carriers), 0),
+           p_nom=ppl.p_nom.where(ppl.carrier.isin(conventional_carriers), 0),
+           p_nom_extendable=ppl.carrier.isin(extendable_carriers['Generator']),
            efficiency=ppl.efficiency,
            marginal_cost=ppl.marginal_cost,
-           capital_cost=0)
+           capital_cost=ppl.capital_cost,
+           build_year=ppl.datein.fillna(0).astype(int),
+           lifetime=(ppl.dateout - ppl.datein).fillna(9999).astype(int),
+        )
+    
+    for carrier in config:
+        
+        # Generators with technology affected
+        idx = n.generators.query("carrier == @carrier").index
+        factors = config[carrier].get("energy_availability_factors")
+
+        if isinstance(v, float):
+            # Single value affecting all generators of technology k indiscriminantely of country
+            n.generators.loc[idx, "p_max_pu"] = v
+        elif isinstance(v, dict):
+            v = pd.Series(v)
+            # Values affecting generators of technology k country-specific
+            # First map generator buses to countries; then map countries to p_max_pu
+            n.generators.p_max_pu.update(n.generators.loc[idx].bus.map(v).dropna())
```

Comment:
updated also the formula, not only variable names. thanks for the description in the comments! unfortunately the formula given in the code didn't do what expected (for me). the following however did:

```suggestion
        if isinstance(factors, float):
            # Single value affecting all generators of technology k indiscriminantely of country
            n.generators.loc[idx, "p_max_pu"] = factors
        elif isinstance(factors, dict):
            factors = pd.Series(factors)
            # Values affecting generators of technology k country-specific
            # First map generator buses to countries; then map countries to p_max_pu
            (n.generators.p_max_pu.update(
                n.generators.loc[idx].bus.map(n.buses.country).dropna().map(factors))
            )
```